cpp/src/arrow/util/list_util.cc

```diff
@@ -0,0 +1,285 @@
+      // valid list-view, but we return the maximum possible value to
+      // avoid undefined behavior.
+      return kInt64Max;
+    }
+  }
+  int64_t max_end =
+      static_cast<int64_t>(offsets[input.offset + i]) + sizes[input.offset + i];
+  if (max_end == input.child_data[0].length) {
+    // Early-exit: maximum possible view-end found already.
+    return max_end;
+  }
+
+  // Slow path: scan the buffers entirely.
+  arrow::internal::VisitSetBitRunsVoid(
+      validity, input.offset, /*length=*/i + 1, [&](int64_t i, int64_t run_length) {
+        for (int64_t j = 0; j < run_length; ++j) {
+          const auto offset = static_cast<int64_t>(offsets[input.offset + i + j]);
+          const auto size = sizes[input.offset + i + j];
+          if (size > 0) {
+            if constexpr (sizeof(offset_type) == sizeof(int64_t)) {
+              if (ARROW_PREDICT_FALSE(offset > kInt64Max - size)) {
+                // 64-bit overflow detected. This is not possible on a valid list-view,
+                // but we saturate max_end to the maximum possible value to avoid
+                // undefined behavior.
+                max_end = kInt64Max;
+                return;
+              }
+            }
+            max_end = std::max(max_end, offset + size);
+          }
+        }
+      });
+  return max_end;
+}
+
+template <typename offset_type>
+std::pair<int64_t, int64_t> RangeOfValuesUsedByListView(const ArraySpan& input) {
+  DCHECK(is_list_view(*input.type));
+  if (input.length == 0 || input.GetNullCount() == input.length) {
+    return {0, 0};
+  }
+  const int64_t min_offset = MinViewOffset<offset_type>(input);
+  const int64_t max_end = MaxViewEnd<offset_type>(input);
+  return {min_offset, max_end - min_offset};
+}
+
+template <typename offset_type>
+std::pair<int64_t, int64_t> RangeOfValuesUsedByList(const ArraySpan& input) {
+  DCHECK(is_var_length_list(*input.type));
+  if (input.length == 0) {
+    return {0, 0};
+  }
+  const auto* offsets = reinterpret_cast<const offset_type*>(input.buffers[1].data);
+  const int64_t min_offset = offsets[input.offset];
+  const int64_t max_end = offsets[input.offset + input.length];
+  return {min_offset, max_end - min_offset};
+}
+
+template <typename DestListViewType, typename SrcListType>
+Result<std::shared_ptr<ArrayData>> ListViewFromListImpl(
+    const std::shared_ptr<ArrayData>& list_data, MemoryPool* pool) {
+  static_assert(
+      std::is_same<typename SrcListType::offset_type,
+                   typename DestListViewType::offset_type>::value,
+      "Offset types between list type and list-view type are expected to match");
+  using offset_type = typename SrcListType::offset_type;
+  const auto& list_type = checked_cast<const SrcListType&>(*list_data->type);
+
+  // To re-use the validity and offsets buffers, a sizes buffer with enough
+  // padding on the beginning is allocated and filled with the sizes after
+  // list_data->offset.
```

Comment:
You must initialize the initial padding area, otherwise you might leak private data accross IPC or the C Data Interface.

Reply:
Adding a `memset` here.